keras_cv/layers/detectron2_layers.py

```diff
@@ -0,0 +1,477 @@
+
+    Args:
+        num_heads (int): Number of attention heads.
+        key_dim (int): Size of each attention head for query, key, and
+            value.
+        use_bias (bool, optional): Whether to use bias when projecting
+            the queries, keys, and values. Defaults to `True`.
+        use_rel_pos (bool, optional): Whether to use relative positional
+            embeddings or not. Defaults to `False`.
+        input_size (tuple[int, int], optional): Size of the input image.
+            Must be provided when using relative positional embeddings.
+            Defaults to `None`.
+
+    Raises:
+        ValueError: When `input_size = None` with `use_rel_pos = True`.
+    """
+
+    def __init__(
+        self,
+        num_heads,
+        key_dim,
+        use_bias=True,
+        use_rel_pos=False,
+        input_size=None,
+        **kwargs
+    ):
+        super().__init__(**kwargs)
+        self.num_heads = num_heads
+        self.key_dim = key_dim
+        self.scale = self.key_dim**-0.5
+        self.use_bias = use_bias
+
+        self.qkv = keras.layers.Dense(
+            key_dim * self.num_heads * 3, use_bias=self.use_bias
+        )
+        self.projection = keras.layers.Dense(key_dim * self.num_heads)
+
+        self.input_size = input_size
+        self.use_rel_pos = use_rel_pos
+        if self.use_rel_pos:
+            if input_size is None:
+                raise ValueError(
+                    "Input size must be provided if using relative "
+                    "positional encoding."
+                )
+            self.rel_pos_h = self.add_weight(
+                name="rel_pos_h",
+                shape=(2 * self.input_size[0] - 1, self.key_dim),
+                initializer="zeros",
+                trainable=True,
+            )
+            self.rel_pos_w = self.add_weight(
+                name="rel_pos_w",
+                shape=(2 * self.input_size[1] - 1, self.key_dim),
+                initializer="zeros",
+                trainable=True,
+            )
+
+    def build(self, input_shape=None):
+        self.qkv.build([self.key_dim * self.num_heads])
+        self.projection.build([self.key_dim * self.num_heads])
+        self.built = True
+
+    def compute_output_shape(self, input_shape):
+        return input_shape
+
+    def call(self, x):
+        shape = ops.shape(x)
+        B, H, W, C = shape[0], shape[1], shape[2], shape[3]
+        qkv = ops.transpose(
+            ops.reshape(
+                self.qkv(x), (B, H * W, 3, self.num_heads, self.key_dim)
+            ),
+            axes=(2, 0, 3, 1, 4),
+        )
+        qkv = ops.reshape(qkv, (3, B * self.num_heads, H * W, self.key_dim))
+        queries, keys, values = ops.unstack(qkv, axis=0)
+        attention_map = (queries * self.scale) @ ops.transpose(
+            keys, axes=(0, 2, 1)
+        )
+
+        if self.use_rel_pos:
+            attention_map = add_decomposed_rel_pos(
```

Comment:
This should probably be a private method as part of a layer

Reply:
Done.